sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/LogicalTypes.java

```diff
@@ -111,4 +134,247 @@ public FieldType getBaseType() {
+     */
+    public EnumerationValue valueOf(int value) {
+      return new EnumerationValue(enumValues.inverse().get(value), value);
+    }
+
+    @Override
+    public String getIdentifier() {
+      return IDENTIFIER;
+    }
+
+    @Override
+    public FieldType getArgumentType() {
+      return FieldType.map(FieldType.STRING, FieldType.INT32);
+    }
+
+    @Override
+    public Map<String, Integer> getArgument() {
+      return enumValues;
+    }
+
+    @Override
+    public FieldType getBaseType() {
+      return FieldType.INT32;
+    }
+
+    @Override
+    public Integer toBaseType(EnumerationValue input) {
+      return input.value;
+    }
+
+    @Override
+    public EnumerationValue toInputType(Integer base) {
+      return valueOf(base);
+    }
+  }
+
+  /**
+   * Represents a single OneOf value. Each object contains an {@link EnumerationValue} specifying
+   * which field is set along with the value of that field.
+   */
+  public static class OneOfValue {
+    private EnumerationValue caseType;
+    private Object value;
+
+    OneOfValue(EnumerationValue caseType, Object value) {
+      this.caseType = caseType;
+      this.value = value;
+    }
+
+    /** Returns the enumeration that specified which OneOf field is set. */
+    public EnumerationValue getCaseType() {
+      return caseType;
+    }
+
+    /** Returns the current value of the OneOf. */
+    @SuppressWarnings("TypeParameterUnusedInFormals")
+    public <T> T getValue() {
+      return (T) value;
+    }
+  }
+  /**
+   * A logical type representing a union of fields. This logical type is initialized with a set of
+   * field and represents a union of those fields. This logical type is backed by a Row object
+   * containing one nullable field matching each input field, and one additional {@link
+   * EnumerationType} logical type field that indicates which field is set.
+   */
+  public static class OneOfType implements LogicalType<OneOfValue, Row> {
+    public static final String IDENTIFIER = "OneOf";
+    public static final String DEFAULT_ENUM_FIELD_NAME = "OneOfCase";
+
+    private final Schema oneOfSchema;
+    private final int oneOfCaseFieldId;
+    private final byte[] schemaProtoRepresentation;
+
+    private OneOfType(List<Field> fields, String enumField) {
+      List<Field> nullableFields =
+          fields.stream()
+              .map(f -> Field.nullable(f.getName(), f.getType()))
+              .collect(Collectors.toList());
+      List<String> enumValues =
+          nullableFields.stream().map(Field::getName).collect(Collectors.toList());
+      nullableFields.add(
+          Field.of(enumField, FieldType.logicalType(EnumerationType.create(enumValues))));
+
+      oneOfSchema = Schema.builder().addFields(nullableFields).build();
+      schemaProtoRepresentation = SchemaTranslation.schemaToProto(oneOfSchema).toByteArray();
+      this.oneOfCaseFieldId = oneOfSchema.indexOf(enumField);
```

Comment:
I'm wondering if the oneOfCase needs to be stored in the row as it can be inferred by the rest of the fields:

- the non-null value is the selected oneOf value
- if all fields are null the oneOf is NULL

this is how proto does it on the wire. Depends on what's important. Data size (extra field) or Case performance.

Reply:
This makes sense. Removed from the row representation